daemon/cluster/convert/swarm.go

```diff
@@ -60,28 +60,36 @@ func SwarmFromGRPC(c swarmapi.Cluster) types.Swarm {
 }
 
 // SwarmSpecToGRPC converts a Spec to a grpc ClusterSpec.
-func SwarmSpecToGRPC(s types.Spec) (swarmapi.ClusterSpec, error) {
-	spec := swarmapi.ClusterSpec{
-		Annotations: swarmapi.Annotations{
-			Name:   s.Name,
-			Labels: s.Labels,
-		},
-		Orchestration: swarmapi.OrchestrationConfig{
-			TaskHistoryRetentionLimit: s.Orchestration.TaskHistoryRetentionLimit,
-		},
-		Raft: swarmapi.RaftConfig{
-			SnapshotInterval:           s.Raft.SnapshotInterval,
-			KeepOldSnapshots:           s.Raft.KeepOldSnapshots,
-			LogEntriesForSlowFollowers: s.Raft.LogEntriesForSlowFollowers,
-			HeartbeatTick:              s.Raft.HeartbeatTick,
-			ElectionTick:               s.Raft.ElectionTick,
-		},
-		Dispatcher: swarmapi.DispatcherConfig{
-			HeartbeatPeriod: ptypes.DurationProto(time.Duration(s.Dispatcher.HeartbeatPeriod)),
-		},
-		CAConfig: swarmapi.CAConfig{
-			NodeCertExpiry: ptypes.DurationProto(s.CAConfig.NodeCertExpiry),
-		},
+func SwarmSpecToGRPC(s types.Spec, initSpec swarmapi.ClusterSpec) (swarmapi.ClusterSpec, error) {
+	spec := initSpec
+
+	// We take the initSpec (either created from scratch, or returned by swarmkit),
+	// and replace the value with the one taken from types.Spec.
+	// In types.Spec we will check the value. If the value is either nil or 0 (depends)
+	// then we will skip the copy as we assume it will be the "default".
```

Comment:
I think it's slightly more accurate to say that if the value is nil or 0 the value won't be changed (maybe it was changed earlier, and it's not the default anymore).